### PR TITLE
Stealing Machine Learning Models via Prediction APIs

### PR DESCRIPTION
**Pull request by AntoineTheb**

### CODE REVIEW

File: article/_posts/2019-11-21-stealml.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+## Online Model Extraction Attacks
+
+The authors then tried to extract models setup online on BigML and Amazon AWS. BigML uses decision tree which will not be adressed in this review. Amazon, however, allows users to upload their data and then Amazon will perform feature extraction and then train a model on these features.
+
+The problem now is to find what features were extracted and which model was used. However, AWS does support "partial input", where you can send an image for prediction which does not have some of the extracted features and still get a result. Without going into details, the authors were able to reverse-engineer the features and still get a copy of the model with 100% matches.
+
+![](/article/images/stealml/tab7.jpeg)
+
+In all the attacks, the authors were able to learn about the model's class, the enable feature extractors and hyperparameters through the ML service AWS's documentation. In a real world scenario, an attacker could narrow down its hyperparameter search through similar techniques and cross-validation.
```
Comment (ppoulin91) [inline article/_posts/2019-11-21-stealml.md]: "the enable feature extractors"... Did you mean enabled?

File: article/_posts/2019-11-21-stealml.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+![](/article/images/stealml/fig4.jpeg)
+![](/article/images/stealml/fig5.jpeg)
+![](/article/images/stealml/fig6.jpeg)
+
+where $$\alpha$$ is the budget, or multiplier of $$(d+1)$$ data points used.
+
+##  Prevention from such attacks
+
+The authors suggest some countermeasures to such attacks, such as rounding off the prediction's accuracy. The results of round the accuracy (or only providing class labels is reported below)
```
Comment (ppoulin91) [inline article/_posts/2019-11-21-stealml.md]: The results of **rounding** the accuracy (or only providing class labels) is reported below.

Comment (AntoineTheb) [reply article/_posts/2019-11-21-stealml.md]: Woops, fixed in 4ef3ccc

File: article/_posts/2019-11-21-stealml.md (hunk 1 above)
Comment (AntoineTheb) [reply article/_posts/2019-11-21-stealml.md]: I did, fixed in 4ef3ccc

File: article/_posts/2019-11-21-stealml.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+cite:
+    authors: "Florian Tramèr, Fan Zhang, Ari Juels, Michael K. Reiter, Thomas Ristenpart"
+    title:   "Stealing Machine Learning Models via Prediction APIs"
+    venue:   "USENIX"
+pdf: "https://arxiv.org/abs/1609.02943"
+---
+
+
+# Highlights
+- Simple equation-solving model extraction attacks for simple ML model
```
Comment (jhlegarreta) [inline article/_posts/2019-11-21-stealml.md]: (...) simple ML model : (...) simple ML models ?

File: article/_posts/2019-11-21-stealml.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+# Introduction
+
+> Machine learning (ML) models may be deemed confidential due to their sensitive training data, commercial value, or use in security applications. Increasingly often, confidential ML models are being deployed with publicly accessible query interfaces. ML-as-a-service (“predictive analytics”) systems are an example: Some charge users for access on a pay-per-query basis.
+> The tension between model confidentiality and public access motivates [the authors'] investigation of model extraction attacks. In such attacks, an adversary with black-box access, but no prior knowledge of an ML model’s parameters or training data, aims to duplicate the functionality of (i.e., “steal”) the model
+
+![](/article/images/stealml/fig1.jpeg)
+
+The goal here is to, after gaining access to a black-box target model $$f$$, train a model $$\hat{f}$$ that closely approximates or even matches $$f$$. Matching here implies that, with $$f_i(x)$$ being either the prediction over x for the _$$i^{th}$$_ class or the confidence score associated with the _$$i^{th}$$_ class, $$f_i(x) \equiv \hat{f}_i(x)$$ $$\geq$$ 99.9% of times.
```
Comment (jhlegarreta) [inline article/_posts/2019-11-21-stealml.md]: (...) the prediction over x for (...) : (...) the prediction over $$x$$ for (...)

File: article/_posts/2019-11-21-stealml.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+# Methods and results
+
+## Data
+
+![](/article/images/stealml/tab3.jpeg)
+
+## Extraction with Confidence Values
+
+First off, extraction attacks on prediction APIs that return confidence values. Here, $$ f(x) \in [0,1]^c$$ where c is the number of classes. For simple enough models, _equation-solving attacks_ can be used to directly infer the model's parameters. For example, in the case of a binary logistic regression (LR) with parameters $$w \in R^d$$ that outputs a probability $$f_1(x) = \sigma(wx)$$, and given a oracle sample $$(x, f(x))$$, we get a linear equation $$wx = \sigma^{-1}(f_1(x))$$ and thus, only d+1 linearly independent samples can be used to recover the parameters.
```
Comment (jhlegarreta) [inline article/_posts/2019-11-21-stealml.md]: (...) where c is the number of classes (...) : (...) where $$c$$ is the number of classes (...)
(...) only d+1 linearly independent (...) : (...) only $$d+1$$ linearly independent (...)

File: article/_posts/2019-11-21-stealml.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+For all the datasets shown in table 3, the authors could train a model and extract it in d+1 predictions with 100% confidence error. For these models, the attack only required 41 queries on average and 113 at most.
```
Comment (jhlegarreta) [inline article/_posts/2019-11-21-stealml.md]: (...) extract it in d+1 predictions (...) : (...) extract it in $$d+1$$ predictions (...)

File: article/_posts/2019-11-21-stealml.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+The problem now is to find what features were extracted and which model was used. However, AWS does support "partial input", where you can send an image for prediction which does not have some of the extracted features and still get a result. Without going into details, the authors were able to reverse-engineer the features and still get a copy of the model with 100% matches.
+
+![](/article/images/stealml/tab7.jpeg)
+
+In all the attacks, the authors were able to learn about the model's class, the enabled feature extractors and hyperparameters through the ML service AWS's documentation. In a real world scenario, an attacker could narrow down its hyperparameter search through similar techniques and cross-validation.
+
+## Extraction Given Class Labels Only
+
+Without going into details, the authors wanted to see if they could extract model through only class predictions. The general strategy is called retraining, where the authors query the models with data points that are close to $$\hat{f}$$'s decision boundary.
```
Comment (jhlegarreta) [inline article/_posts/2019-11-21-stealml.md]: (...) they could extract model (...) : (...) they could extract a (the?) model (...)

File: article/_posts/2019-11-21-stealml.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+![](/article/images/stealml/fig4.jpeg)
+![](/article/images/stealml/fig5.jpeg)
+![](/article/images/stealml/fig6.jpeg)
+
+where $$\alpha$$ is the budget, or multiplier of $$(d+1)$$ data points used.
+
+##  Prevention from such attacks
+
+The authors suggest some countermeasures to such attacks, such as rounding off the prediction's accuracy. The impacts of rounding the accuracy (or only providing class labels) is reported below
```
Comment (jhlegarreta) [inline article/_posts/2019-11-21-stealml.md]: The impacts (...) are reported below.
or
The impact (...) is reported below.

File: article/_posts/2019-11-21-stealml.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+![](/article/images/stealml/fig7.jpeg)
+
+They also mention using ensemble methods, as recovering $$n$$ times model parameters is much more costly and only one set of model parameters.
+
+# Conclusions
+> [The authors] demonstrated how the flexible prediction APIs exposed by current ML-as-a-service providers enable new model extraction attacks that could subvert model monetization, violate training-data privacy, and facilitate model evasion.
+
+# Remarks
+
+- All the attacks were done on models setup by the authors themselves so as not to potentially harm anyone.
```
Comment (jhlegarreta) [inline article/_posts/2019-11-21-stealml.md]: *setup* is the noun; *set up* is the verb if I'm not mistaken

File: article/_posts/2019-11-21-stealml.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+- The authors also attacked decision tree models and Kernel SVMs with success. They also performed _model inversion_ attacks to recover training data with variable success.
+- All the attacks presumed the attackers knows the models' architectures
```
Comment (jhlegarreta) [inline article/_posts/2019-11-21-stealml.md]: (...) the attackers knows (...) : (...) the attacker knows (...) or (...) the attackers know (...) architectures.

File: article/_posts/2019-11-21-stealml.md [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+- All the attacks are heavily dependant on the number of parameters and, therefore, modern models might prove to be too costly for these attacks.
```
Comment (jhlegarreta) [inline article/_posts/2019-11-21-stealml.md]: dependant : dependent

Comment (AntoineTheb) [reply article/_posts/2019-11-21-stealml.md]: _dependant_ is the Canadian english spelling according to the very serious
http://www.lukemastin.com/testing/spelling/cgi-bin/database.cgi?action=view_category&database=spelling&category=D

File: article/_posts/2019-11-21-stealml.md (hunk 10 above)
Comment (AntoineTheb) [reply article/_posts/2019-11-21-stealml.md]: fixed in a5fac43

File: article/_posts/2019-11-21-stealml.md (hunk 9 above)
Comment (AntoineTheb) [reply article/_posts/2019-11-21-stealml.md]: fixed in a5fac43

File: article/_posts/2019-11-21-stealml.md (hunk 8 above)
Comment (AntoineTheb) [reply article/_posts/2019-11-21-stealml.md]: fixed in a5fac43

File: article/_posts/2019-11-21-stealml.md (hunk 7 above)
Comment (AntoineTheb) [reply article/_posts/2019-11-21-stealml.md]: fixed in a5fac43

File: article/_posts/2019-11-21-stealml.md (hunk 6 above)
Comment (AntoineTheb) [reply article/_posts/2019-11-21-stealml.md]: fixed in a5fac43

File: article/_posts/2019-11-21-stealml.md (hunk 5 above)
Comment (AntoineTheb) [reply article/_posts/2019-11-21-stealml.md]: fixed in a5fac43

File: article/_posts/2019-11-21-stealml.md (hunk 4 above)
Comment (AntoineTheb) [reply article/_posts/2019-11-21-stealml.md]: fixed in a5fac43

File: article/_posts/2019-11-21-stealml.md (hunk 3 above)
Comment (AntoineTheb) [reply article/_posts/2019-11-21-stealml.md]: fixed in a5fac43

File: article/_posts/2019-11-21-stealml.md (hunk 11 above)
Comment (jhlegarreta) [reply article/_posts/2019-11-21-stealml.md]: :+1: Didn't know that there was yet another spelling for that. Sorry !

**Pull request merged**